TST: Add test to ensure DF describe does not throw an error (#32409)

luckyvs1 opened the pull request:
- [x] closes #32409
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Notes:
- Tested using `pytest pandas/tests/frame/methods/test_describe.py::TestDataFrameDescribe::test_describe_does_not_raise_error` and tested error format by calling a raise error line in the try block `raise TypeError("Test message")`

luckyvs1 committed:
TST: Add test to ensure Dataframe describe does not throw an error (#32409)

luckyvs1 committed:
TST: Update linting with black pandas formatting

jreback commented on pandas/tests/frame/methods/test_describe.py:
```diff
@@ -298,3 +299,23 @@ def test_describe_percentiles_integer_idx(self):
+        expected = DataFrame(
+            {"test": [2, 2, {"a": "1"}, 1]}, index=["count", "unique", "top", "freq"]
+        )
+        try:
```
you don't need the try/except if it fails then pytest automatically raises.

luckyvs1 replied:
Thank you for the info, I've updated this PR and re-requested a review

luckyvs1 committed:
TST: Remove try except to catch error and fail test

luckyvs1 committed:
TST: Remove unused pytest import

TomAugspurger commented:
This is very strange, but I  don't know that this test is actually covering the original issue. With pandas 1.0.1

```python
In [6]: >>> df = pd.DataFrame([{"test": {"a": "1"}}, {"test": {"a": "2"}}])

In [7]: r = df.describe()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
pandas/_libs/hashtable_class_helper.pxi in pandas._libs.hashtable.PyObjectHashTable.map_locations()

TypeError: unhashable type: 'dict'
Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas/_libs/hashtable_class_helper.pxi", line 1652, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'dict'

In [8]: r
Out[8]:
              test
count            2
unique           2
top     {'a': '1'}
freq             1
```

So the code executed. The `TypeError` printed there is apparently handled, but it's still printed out to the screen, and the traceback has been set.

@luckyvs1 can you verify that the test you added *fails* with pandas 1.0.1? You should be able just make a single `test_.py` file with your new test in it (make sure it's not in your `pandas` directory). And then create a new environment with pandas 1.0.1 and pytest and run the test?

luckyvs1 commented:
Thank you for the feedback @TomAugspurger -- unfortunately I won't be able to look into this till this weekend, but will investigate if the test case written fails in the tagged `1.0.1` release version and if it doesn’t will modify it accordingly and retest it passes on the latest.

luckyvs1 commented:
@TomAugspurger  I just had a chance to verify on the release `1.0.1` [branch](https://github.com/luckyvs1/pandas/tree/add-test-for-describe-v1.0.1) this test does indeed throw an error:

```
ubuntu@ubuntu-MacBookPro:~/Desktop/Github/pandas$ pytest pandas/tests/frame/methods/test_describe.py::TestDataFrameDescribe::test_describe_does_not_raise_error
================================================================= test session starts ==================================================================
platform linux -- Python 3.6.9, pytest-5.4.3, py-1.9.0, pluggy-0.13.1
rootdir: /home/ubuntu/Desktop/Github/pandas, inifile: setup.cfg
plugins: hypothesis-5.19.0, cov-2.10.0, responses-0.4.0
collected 1 item                                                                                                                                       

pandas/tests/frame/methods/test_describe.py F                                                                                                    [100%]

======================================================================= FAILURES =======================================================================
_______________________________________________ TestDataFrameDescribe.test_describe_does_not_raise_error _______________________________________________

self = <pandas.tests.frame.methods.test_describe.TestDataFrameDescribe object at 0x7f28f1cf51d0>

    def test_describe_does_not_raise_error(self):
        # GH#32409
        df = pd.DataFrame([{"test": {"a": "1"}}, {"test": {"a": "2"}}])
        expected = DataFrame(
            {"test": [2, 2, {"a": "1"}, 1]}, index=["count", "unique", "top", "freq"]
        )
>       result = df.describe()

pandas/tests/frame/methods/test_describe.py:340: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pandas/core/generic.py:9940: in describe
    ldesc = [describe_1d(s) for _, s in data.items()]
pandas/core/generic.py:9940: in <listcomp>
    ldesc = [describe_1d(s) for _, s in data.items()]
pandas/core/generic.py:9923: in describe_1d
    return describe_categorical_1d(data)
pandas/core/generic.py:9879: in describe_categorical_1d
    objcounts = data.value_counts()
pandas/core/base.py:1244: in value_counts
    dropna=dropna,
pandas/core/algorithms.py:723: in value_counts
    keys, counts = _value_counts_arraylike(values, dropna)
pandas/core/algorithms.py:767: in _value_counts_arraylike
    keys, counts = f(values, dropna)
pandas/_libs/hashtable_func_helper.pxi:336: in pandas._libs.hashtable.value_count_object
    cpdef value_count_object(ndarray[object] values, bint dropna):
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   result_keys = np.empty(table.n_occupied, dtype=np.object)
E   TypeError: unhashable type: 'dict'

pandas/_libs/hashtable_func_helper.pxi:347: TypeError
=============================================================== short test summary info ================================================================
FAILED pandas/tests/frame/methods/test_describe.py::TestDataFrameDescribe::test_describe_does_not_raise_error - TypeError: unhashable type: 'dict'
================================================================== 1 failed in 0.39s ===================================================================
```

and the output of the test on my current PR branch which is built off release `1.0.5`:

```
ubuntu@ubuntu-MacBookPro:~/Desktop/Github/pandas$ pytest pandas/tests/frame/methods/test_describe.py::TestDataFrameDescribe::test_describe_does_not_raise_error
================================================================= test session starts ==================================================================
platform linux -- Python 3.6.9, pytest-5.4.3, py-1.9.0, pluggy-0.13.1
rootdir: /home/ubuntu/Desktop/Github/pandas, inifile: setup.cfg
plugins: hypothesis-5.19.0, cov-2.10.0, responses-0.4.0
collected 1 item                                                                                                                                       

pandas/tests/frame/methods/test_describe.py .                                                                                                    [100%]

================================================================== 1 passed in 0.04s ===================================================================
```

simonjayhawkins commented:
something weird going on. On windows with pandas 1.0.1, I get

```
$ pytest ~/test.py -vv -s
============================= test session starts =============================
platform win32 -- Python 3.8.3, pytest-5.4.3, py-1.9.0, pluggy-0.13.1 -- C:\Users\simon\Anaconda3\envs\pandas-1.0.1\python.exe
cachedir: .pytest_cache
rootdir: C:\Users\simon
collecting ... collected 1 item

..\test.py::test_describe_does_not_raise_error Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 1652, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'dict'
Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 1652, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'dict'
PASSED

============================== 1 passed in 0.79s ==============================
```

luckyvs1 commented:
@simonjayhawkins  Using `Python 3.6.9` with `Pandas 1.0.1` on `Linux Ubuntu 18.04.4 LTS` this issue doesn't seem to be resolved:

I imported `pandas 1.0.1` in the Python shell and performed the example as in the issue and the describe command indeed did not run as seen in the result:

## Reproduced example with failed describe in shell with `pandas 1.0.1`
```
ubuntu@ubuntu-MacBookPro:~/Desktop/Github$ python3
Python 3.6.9 (default, Apr 18 2020, 01:56:04) 
[GCC 8.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pandas as pd
>>> import numpy as np
>>> print(pd.__version__)
1.0.1
>>> df = pd.DataFrame([{"test": {"a": "1"}}, {"test": {"a": "2"}}])
>>> print(df)
         test
0  {'a': '1'}
1  {'a': '2'}
>>> print(df.describe())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ubuntu/.local/lib/python3.6/site-packages/pandas/core/generic.py", line 9940, in describe
    ldesc = [describe_1d(s) for _, s in data.items()]
  File "/home/ubuntu/.local/lib/python3.6/site-packages/pandas/core/generic.py", line 9940, in <listcomp>
    ldesc = [describe_1d(s) for _, s in data.items()]
  File "/home/ubuntu/.local/lib/python3.6/site-packages/pandas/core/generic.py", line 9923, in describe_1d
    return describe_categorical_1d(data)
  File "/home/ubuntu/.local/lib/python3.6/site-packages/pandas/core/generic.py", line 9879, in describe_categorical_1d
    objcounts = data.value_counts()
  File "/home/ubuntu/.local/lib/python3.6/site-packages/pandas/core/base.py", line 1244, in value_counts
    dropna=dropna,
  File "/home/ubuntu/.local/lib/python3.6/site-packages/pandas/core/algorithms.py", line 723, in value_counts
    keys, counts = _value_counts_arraylike(values, dropna)
  File "/home/ubuntu/.local/lib/python3.6/site-packages/pandas/core/algorithms.py", line 767, in _value_counts_arraylike
    keys, counts = f(values, dropna)
  File "pandas/_libs/hashtable_func_helper.pxi", line 336, in pandas._libs.hashtable.value_count_object
  File "pandas/_libs/hashtable_func_helper.pxi", line 347, in pandas._libs.hashtable.value_count_object
TypeError: unhashable type: 'dict'
>>> result = df.describe()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ubuntu/.local/lib/python3.6/site-packages/pandas/core/generic.py", line 9940, in describe
    ldesc = [describe_1d(s) for _, s in data.items()]
  File "/home/ubuntu/.local/lib/python3.6/site-packages/pandas/core/generic.py", line 9940, in <listcomp>
    ldesc = [describe_1d(s) for _, s in data.items()]
  File "/home/ubuntu/.local/lib/python3.6/site-packages/pandas/core/generic.py", line 9923, in describe_1d
    return describe_categorical_1d(data)
  File "/home/ubuntu/.local/lib/python3.6/site-packages/pandas/core/generic.py", line 9879, in describe_categorical_1d
    objcounts = data.value_counts()
  File "/home/ubuntu/.local/lib/python3.6/site-packages/pandas/core/base.py", line 1244, in value_counts
    dropna=dropna,
  File "/home/ubuntu/.local/lib/python3.6/site-packages/pandas/core/algorithms.py", line 723, in value_counts
    keys, counts = _value_counts_arraylike(values, dropna)
  File "/home/ubuntu/.local/lib/python3.6/site-packages/pandas/core/algorithms.py", line 767, in _value_counts_arraylike
    keys, counts = f(values, dropna)
  File "pandas/_libs/hashtable_func_helper.pxi", line 336, in pandas._libs.hashtable.value_count_object
  File "pandas/_libs/hashtable_func_helper.pxi", line 347, in pandas._libs.hashtable.value_count_object
TypeError: unhashable type: 'dict'
>>> print(result)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'result' is not defined
```

## Ubuntu Version
```
ubuntu@ubuntu-MacBookPro:~/Desktop/Github$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 18.04.4 LTS
Release:	18.04
Codename:	bionic
```

## Output of pd.show_versions()
```
>>> pd.show_versions()

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.6.9.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.15.0-111-generic
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.0.1
numpy            : 1.19.0
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 9.0.1
setuptools       : 49.2.0
Cython           : 0.29.21
pytest           : 5.4.3
hypothesis       : 5.19.0
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.4.2
html5lib         : 0.999999999
pymysql          : None
psycopg2         : None
jinja2           : 2.10.3
IPython          : 7.10.1
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : 4.4.2
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pytest           : 5.4.3
pyxlsb           : None
s3fs             : None
scipy            : 1.5.1
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None
numba            : None
```

Please let me know if you need any additional information.

simonjayhawkins commented:
I'm not sure how this should be tested (or whether it can be tested). my concern is that the original cause of the issue of not catching exceptions in the cython code is not solved (see https://github.com/pandas-dev/pandas/pull/34113#issue-416019180). I'm not sure how the issue is fixed on master. Maybe I did the git bisect wrong in https://github.com/pandas-dev/pandas/issues/32409#issuecomment-606627286. The pr referenced there changed is_scalar and is_iterator, so if that PR did fix the describe issue, it must have changed the behaviour of one of those functions. Either way, I think there is a problem somewhere.

luckyvs1 commented:
For reference I'm also able to reproduce the issue described in #32409 on my Mac as well.

## Reproduced example:
```
Luckys-MacBook-Pro:~ lucky$ python3
Python 3.6.5 | packaged by conda-forge | (default, Apr  6 2018, 13:44:09) 
[GCC 4.2.1 Compatible Apple LLVM 6.1.0 (clang-602.0.53)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import pandas as pd
>>> import numpy as np
>>> print(pd.__version__)
1.0.1
>>> df = pd.DataFrame([{"test": {"a": "1"}}, {"test": {"a": "2"}}])
>>> print(df)
         test
0  {'a': '1'}
1  {'a': '2'}
>>> print(df.describe())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/lucky/anaconda/lib/python3.6/site-packages/pandas/core/generic.py", line 9940, in describe
    ldesc = [describe_1d(s) for _, s in data.items()]
  File "/Users/lucky/anaconda/lib/python3.6/site-packages/pandas/core/generic.py", line 9940, in <listcomp>
    ldesc = [describe_1d(s) for _, s in data.items()]
  File "/Users/lucky/anaconda/lib/python3.6/site-packages/pandas/core/generic.py", line 9923, in describe_1d
    return describe_categorical_1d(data)
  File "/Users/lucky/anaconda/lib/python3.6/site-packages/pandas/core/generic.py", line 9879, in describe_categorical_1d
    objcounts = data.value_counts()
  File "/Users/lucky/anaconda/lib/python3.6/site-packages/pandas/core/base.py", line 1244, in value_counts
    dropna=dropna,
  File "/Users/lucky/anaconda/lib/python3.6/site-packages/pandas/core/algorithms.py", line 723, in value_counts
    keys, counts = _value_counts_arraylike(values, dropna)
  File "/Users/lucky/anaconda/lib/python3.6/site-packages/pandas/core/algorithms.py", line 767, in _value_counts_arraylike
    keys, counts = f(values, dropna)
  File "pandas/_libs/hashtable_func_helper.pxi", line 336, in pandas._libs.hashtable.value_count_object
  File "pandas/_libs/hashtable_func_helper.pxi", line 347, in pandas._libs.hashtable.value_count_object
TypeError: unhashable type: 'dict'
>>> result = df.describe()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/lucky/anaconda/lib/python3.6/site-packages/pandas/core/generic.py", line 9940, in describe
    ldesc = [describe_1d(s) for _, s in data.items()]
  File "/Users/lucky/anaconda/lib/python3.6/site-packages/pandas/core/generic.py", line 9940, in <listcomp>
    ldesc = [describe_1d(s) for _, s in data.items()]
  File "/Users/lucky/anaconda/lib/python3.6/site-packages/pandas/core/generic.py", line 9923, in describe_1d
    return describe_categorical_1d(data)
  File "/Users/lucky/anaconda/lib/python3.6/site-packages/pandas/core/generic.py", line 9879, in describe_categorical_1d
    objcounts = data.value_counts()
  File "/Users/lucky/anaconda/lib/python3.6/site-packages/pandas/core/base.py", line 1244, in value_counts
    dropna=dropna,
  File "/Users/lucky/anaconda/lib/python3.6/site-packages/pandas/core/algorithms.py", line 723, in value_counts
    keys, counts = _value_counts_arraylike(values, dropna)
  File "/Users/lucky/anaconda/lib/python3.6/site-packages/pandas/core/algorithms.py", line 767, in _value_counts_arraylike
    keys, counts = f(values, dropna)
  File "pandas/_libs/hashtable_func_helper.pxi", line 336, in pandas._libs.hashtable.value_count_object
  File "pandas/_libs/hashtable_func_helper.pxi", line 347, in pandas._libs.hashtable.value_count_object
TypeError: unhashable type: 'dict'
>>> print(result)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'result' is not defined
```

## Mac Version
```
Luckys-MacBook-Pro:~ lucky$ sw_vers
ProductName:	Mac OS X
ProductVersion:	10.15.4
BuildVersion:	19E287
```

## Output of pd.show_versions()
```

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.6.5.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 19.4.0
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_CA.UTF-8
LOCALE           : en_CA.UTF-8

pandas           : 1.0.1
numpy            : 1.17.5
pytz             : 2017.2
dateutil         : 2.8.1
pip              : 20.1.1
setuptools       : 49.1.0.post20200704
Cython           : 0.25.2
pytest           : 5.4.3
hypothesis       : 5.19.0
sphinx           : 1.5.6
blosc            : None
feather          : None
xlsxwriter       : 0.9.6
lxml.etree       : 3.7.3
html5lib         : 0.999
pymysql          : None
psycopg2         : None
jinja2           : 2.9.6
IPython          : 5.3.0
pandas_datareader: None
bs4              : 4.6.0
bottleneck       : 1.3.2
fastparquet      : None
gcsfs            : None
lxml.etree       : 3.7.3
matplotlib       : 3.2.2
numexpr          : 2.7.1
odfpy            : None
openpyxl         : 2.4.7
pandas_gbq       : None
pyarrow          : None
pytables         : None
pytest           : 5.4.3
pyxlsb           : None
s3fs             : None
scipy            : 1.5.0
sqlalchemy       : 1.1.9
tables           : 3.6.1
tabulate         : None
xarray           : None
xlrd             : 1.0.0
xlwt             : 1.2.0
xlsxwriter       : 0.9.6
numba            : 0.49.1
```

luckyvs1 commented:
If this needs to be fixed in 1.0.1.x then this test or similar can be added to that release and check if the added test passes on Mac, Linux versions supported in addition to Windows. Since the current behaviour seems to be it's working on Windows, but not Mac and Linux (at least the versions I provided).

In the versions where the exception is handled while still being outputted like the current Windows 1.0.1, I’m also not sure at the moment on how that can be tested or how to currently test the lack of exception output as seen in latest 1.0.5+ versions.

MicaelJarniac committed:
DOC: Fix typo (#37980)

fangchenli committed:
BLD: set inplace in setup.cfg (#37973)

jbrockmendel committed:
DEPR: how keyword in PeriodIndex.astype (#37982)

ma3da committed:
BUG: DataFrame.to_html ignores formatters (#37958)

nrebena committed:
Inconsistent indexes for tick label plotting (#28733)

* TST: Test for issues #26186 and #11465

* BUG: Generate the tick position in BarPlot using convert tools from matlab.

Generate the tick position in BarPlot using convert tools from matlab.

* TST: Modify tests/plotting/test_frame.test_bar_categorical

Ticklocs are now float also for categorical bar data (as they are
position on the axis). The test is changed to compare to a array of
np.float.

* TST: Fix test for windows OS

* TST: Add test for plotting MultiIndex bar plot

A fix to issue #26186 revealed no tests existed about plotting a bar
plot for a MultiIndex, but a section of the user guide visualization
did. This section of the user guide is now in the test suite.

* BUG: Special case for MultiIndex bar plot

* DOC: Add whatsnew entry for PR #28733

* CLN: Clean up in code and doc

* CLN: Clean up test_bar_numeric

* DOC Move to whatsnew v1.1

* FIX: Make tick dtype int for backwards compatibility

* DOC: Improve whatsnew message

* ENH: Add UserWarning when plotting bar plot with MultiIndex

* CLN: Remove duplicate code line

* TST: Capture UserWarning for Bar plot with MultiIndex

* TST: Improve test explanation

* ENH: Raise UserWarning only if redrawing on existing axis with data

* DOC: Move to whatsnew v1.2.9

Co-authored-by: Marco Gorelli <m.e.gorelli@gmail.com>

phofl committed:
BUG: Bug in setitem raising ValueError when setting more than one column via array (#37964)

jbrockmendel committed:
CLN: make MultiIndex._shallow_copy signature match other subclasses (#37985)

jbrockmendel committed:
PERF: IntervalArray.argsort (#37971)

phofl committed:
Bug in loc did not raise KeyError when missing combination with slice(None) was given (#37764)

arw2019 committed:
TST: add nullable array frame constructor dtype tests (#37972)

realead committed:
PERF: Introducing HashTables for datatypes with 8,16 and 32 bits (#37920)

phofl committed:
BUG: Fix inconsistent ordering between left and right in merge (#37406)

jbrockmendel committed:
BUG: CategoricalIndex.where nulling out non-categories (#37977)

* BUG: CategoricalIndex.where nulling out non-categories

* whatsnew

jbrockmendel committed:
BUG: isin incorrectly casting ints to datetimes (#37528)

* BUG: isin incorrectly casting ints to datetimes

* GH ref

* add asvs

jbrockmendel committed:
BUG: IntervalArray.astype(categorical_dtype) losing ordered (#37984)

* BUG: IntervalArray.astype(categorical_dtype) losing ordered

* whatsnew

jbrockmendel committed:
TST/REF: collect indexing tests by method (#38005)

jbrockmendel committed:
REF: ensure_arraylike in algos.isin (#38004)

attack68 committed:
DOC: add a link to new styler method (#37998)

jbrockmendel committed:
TST/REF: collect tests from test_multilevel (#38006)

jbrockmendel committed:
CLN: avoid try/except in Index methods (#37990)

fangchenli committed:
CLN: remove panel compat shim (#37983)

xinrong-meng committed:
[WIP] DOC: MultiIndex EX01 errors (#37993)

fgebhart committed:
TST: add test to verify column does not lose categorical type when using loc (#37988)

arw2019 committed:
REGR: fix inplace operations for EAs with non-EA arg (#37986)

jbrockmendel committed:
REF: Implement isin on DTA instead of DTI (#38012)

fangchenli committed:
CLN: fix E741 ambiguous variable #34150 (#38009)

jbrockmendel committed:
CLN: always pass ndim to make_block (#37991)

jbrockmendel committed:
CLN: Use new hashtables in libindex to avoid casting (#37994)

vineethraj510 committed:
Removed  period to make it more consisent (#37995)

jbrockmendel committed:
TST/REF: collect Index.equals tests (#38011)

jbrockmendel committed:
TST/REF: collect Index setops tests (#38019)

rhshadrach committed:
TYP: Add cast to ABC classes. (#37902)

arw2019 committed:
TST: Series construction from ExtensionDtype scalar (#37989)

ivanovmg committed:
CLN/TYP: pandas/io/formats/excel.py (#37862)

jbrockmendel committed:
REF: define DTA._infer_matches (#38008)

attack68 committed:
ENH: Styler column and row styles (#35607)

ivanovmg committed:
REF: prepare dataframe info for series info (#37868)

jorisvandenbossche committed:
Revert "REF: back IntervalArray by a single ndarray (#37047)" (#38024)

This reverts commit 9cb372376fc78aa66e2559de919592007b74cfaa.

phofl committed:
Bug in loc raised Error when non-integer slice was given for MultiIndex (#37707)

phofl committed:
Bug in loc raising KeyError when MultiIndex columns has only one level (#38026)

* Bug in loc raising KeyError when MultiIndex columns has only one level

* Move test

* Revert commit

jbrockmendel committed:
BUG: pytables in py39 (#38041)

* BUG: pytables in py39

* add pytables to py39 ci

* whatsnew 1.1.5

* scipy and pyarrow to py39 build

Abo7atm committed:
DOC: add unqiue value counts example to groupyby guide (#37847)

phofl committed:
Bug in DataFrame.loc returning elements in wrong order when indexer is differently ordered than object (#37992)

TomAugspurger committed:
Implement DataFrame.__array_ufunc__ (#36955)

hongshaoyang committed:
DOC: Deprecate null_counts parameter of DataFrame.info (#37999)

jbrockmendel committed:
TST/REF: collect Index formatting tests (#38050)

jbrockmendel committed:
BUG: algos.isin numeric vs datetimelike (#38020)

rhshadrach committed:
DOC: Whatsnew 1.2.0 cleanup (#38002)

ivanovmg committed:
TST: parametrize test_info (#37887)

jbrockmendel committed:
CLN: remove ABCIndex (#38055)

phofl committed:
ENH: Implement cross method for Merge Operations (#37864)

phofl committed:
BUG: __getitem__ raise blank KeyError for IntervalIndex and missing keys (#37873)

simonjayhawkins committed:
DOC: tidy 1.1.5 release notes (#38064)

Co-authored-by: Joris Van den Bossche <jorisvandenbossche@gmail.com>

lidavidm committed:
PERF: fix regression in tz_convert_from_utc (#38074)

phofl committed:
ENH: Add support for min_count keyword for Resample and Groupby functions (#37870)

jorisvandenbossche committed:
DOC: clarify docs related to deprecation of indexing DataFrame with single partial datetime-string (#38088)

steveya committed:
BUG: pandas.DataFrame().stack() raise an error, while expected is empty (#36185)

jbrockmendel committed:
CLN: inconsistent namespace usage in tests.indexes (#38076)

rhshadrach committed:
CLN: Remove .values from groupby.sem (#38044)

topper-123 committed:
DOC: Clean .ix references in indexing.rst (#38080)

jbrockmendel committed:
REF: Share code between NumericIndex subclasses (#38027)

arw2019 committed:
CLN/TST: delegate StringArray.fillna() to parent class + add tests (#37987)

jorisvandenbossche committed:
REGR: revert "CLN: _consolidate_inplace less" / fix regression in fillna() (#34407)

jbrockmendel committed:
BUG: NumericIndex.insert(0, False) casting to int (#38030)

jbrockmendel committed:
CLN: dont mix Int64Index into PeriodIndex (#38045)

eumiro committed:
DOC: Remove repeated words and wrong /it's/ usage (#38047)

realead committed:
ENH: adding support for Py3.6+ memory tracing for khash-maps (#38048)

fangchenli committed:
CLN: fix flake8 C406, C409, and some of C408 (#38078)

jbrockmendel committed:
REF: share argmax axis validation test across all indexes (#38049)

jbrockmendel committed:
BUG: get_indexer_non_unique not excluding boolean like get_indexer (#38052)

jbrockmendel committed:
TST/REF: take things out of Base tests (#38072)

ma3da committed:
TST: DataFrame.at on categorical with missing (#38085)

jreback commented:
@luckyvs1 this got lost, can you merge master and ping on green

xhochy committed:
BLD/DOC: Use conda(-forge) compilers (#38084)

jbrockmendel committed:
ENH: preserve RangeIndex in factorize (#38034)

* ENH: preserve RangeIndex in factorize

* dedoc

* ensure np.intp

* 32bit compat

* if->elif

jbrockmendel committed:
BUG: raise consistent exception on slicing failure (#38077)

jbrockmendel committed:
BUG: loc.setitem corner case (#37931)

* BUG: loc.setitem corner case

* flesh out comment

jbrockmendel committed:
BUG: RangeIndex.difference with mismatched step (#38028)

* BUG: return RangeIndex from RangeIndex.difference

* test coverage

* BUG: RangeIndex.difference with mismatched step

* whatsnew

GYHHAHA committed:
BUG: MultiIndex.drop does not raise if labels are partially found (#37830)

jbrockmendel committed:
TST: tighten assert_index_equal calls (#38054)

mroeschke committed:
CLN: pd. namespace in tests/window (#38095)

Co-authored-by: Matt Roeschke <mroeschke@housecanary.com>

phofl committed:
[TST]: Add test for duplicate keys in concat (#36556)

* Add test for 20816

* Modify test

* Move test

jbrockmendel committed:
PERF: replace_list (#38097)

jbrockmendel committed:
PERF: Index.searchsorted (#38103)

jbrockmendel committed:
BUG: loc.setitem with expansion expanding rows (#37932)

* BUG: loc.setitem with expansion expanding rows

* split test

* whatsnew

xhochy committed:
BLD: Only enable -Werror in the CI jobs (#38087)

Co-authored-by: Simon Hawkins <simonjayhawkins@gmail.com>

jorisvandenbossche committed:
REGR: fix regression in groupby aggregation with out-of-bounds datetimes (#38094)

jorisvandenbossche committed:
TST: rewrite convert_dtypes test to make it easier extendable (#38113)

* TST: rewrite convert_dtypes test to make it easier extendable

* formatting

arw2019 committed:
BUG: crosstab with duplicate column or index labels (#37997)

mroeschke committed:
TST: Suppress http logs in tests (#38096)

fangchenli committed:
CLN: fix flake8 C408 part 2 (#38116)

ivanovmg committed:
TST: mark all plotting tests as slow (#38079)

jbrockmendel committed:
CLN: remove unreachable in core.dtypes (#38128)

jbrockmendel committed:
CLN: simplify mask_missing (#38127)

rhshadrach committed:
TYP: Add cast to ABC Index-like types (#38043)

simonjayhawkins committed:
DOC: add contibutors to 1.2.0 release notes (#38132)

jbrockmendel committed:
REF: IntervalIndex._assert_can_do_setop (#38112)

phofl committed:
BUG: Series.at returning Series with one element instead of scalar (#38101)

jbrockmendel committed:
API: CategoricalIndex.append fallback to concat_compat (#38098)

mlondschien committed:
BUG: fix astype conversion string -> float (#37974)

jorisvandenbossche committed:
ENH: add use_nullable_dtypes option in read_parquet (#31242)

phofl committed:
BUG: Fix duplicates in intersection of multiindexes (#36927)

GYHHAHA committed:
BUG: merge_ordered fails with list-like left_by or right_by (#38089)

hongshaoyang committed:
DOC: Add behavior for Index argument in DataFrame.loc (#38109)

jbrockmendel committed:
REF: de-duplicate ndarray[datetimelike] wrapping (#38129)

jbrockmendel committed:
BUG: Index.intersection casting to object instead of numeric (#38122)

jbrockmendel committed:
CLN: remove unnecesary cast.maybe_convert_objects (#38144)

topper-123 committed:
API: membership checks on ExtensionArray containing NA values (#37867)

jorisvandenbossche committed:
ENH: include conversion to nullable float in convert_dtypes() (#38117)

jbrockmendel committed:
DEPR: ExtensionOpsMixin -> OpsMixin (#38142)

jbrockmendel committed:
REF: use np.where instead of maybe_upcast_putmask in nanops (#38130)

jbrockmendel committed:
CLN: remove unreachable in maybe_cast_result (#38152)

phofl committed:
ENH: Improve performance for df.__setitem__ with list-like indexers (#38148)

jbrockmendel committed:
CLN: remove unused coerce arg in NDFrame._convert (#38151)

jbrockmendel committed:
BUG: Categorical[dt64tz].to_numpy() losing tz (#38136)

* BUG: Categorical[dt64tz].to_numpy() losing tz

* Update pandas/core/arrays/categorical.py

Co-authored-by: gfyoung <gfyoung17+GitHub@gmail.com>

* Avoid FutureWarning

* whatsnew

Co-authored-by: gfyoung <gfyoung17+GitHub@gmail.com>

jorisvandenbossche committed:
Revert "DEPR: ExtensionOpsMixin -> OpsMixin (#38142)" (#38158)

This reverts commit f65f0d3edb275faf37435ba1fa2780240b105b48.

jorisvandenbossche committed:
DOC: trim error traceback in allows_duplicate_labels whatsnew entry (#38184)

jbrockmendel committed:
CLN: remove _recast_datetimelike_result (#38156)

jorisvandenbossche committed:
Revert "REF: define nanargminmax without values_for_argsort (#37815)" (#38177)

This reverts commit 840c1425541917ae820660cd66f6b8b5f4db2387.

arw2019 committed:
BUG: combine_first does not retain dtypes with Timestamp DataFrames (#38145)

jbrockmendel committed:
API: preserve freq in DTI/TDI.factorize (#38120)

jbrockmendel committed:
REF: remove unnecessary maybe_cast_result (#38165)

jbrockmendel committed:
REF: more distinctive names for BaseGrouper.aggregate, transform (#38161)

skvrahul committed:
Rewrite dict literal for files in tests/reshape/ (#38180)

liudj2008 committed:
fix frame, generic, series.py (#38169)

jbrockmendel committed:
CI: update tests for numpy 1.20 change to floordiv (#38172)

Qbiwan committed:
change dict to literal (#38186)

Qbiwan committed:
change dict to literal (#38188)

jorisvandenbossche committed:
BUG: fix wrong error message in deprecated 2D indexing of Series with datetime values (#38099)

twoertwein committed:
BUG/CLN: Minimize number of ResourceWarnings (#38168)

jorisvandenbossche committed:
PERF: fix regression in creation of resulting index in RollingGroupby (#38057)

jorisvandenbossche committed:
BUG: preserve nullable dtype for float result in IntegerArray/BooleanArray arithmetic ops (#38178)

simonjayhawkins committed:
Revert "ENH: Improve performance for df.__setitem__ with list-like indexers (#38148)" (#38208)

This reverts commit 2f4110967c303c739abef67be63ba2206d64bbb9.

jorisvandenbossche committed:
CI/TST: fix CI with numpy dev for changed error message / dev version (#38209)

jbrockmendel committed:
CI: xfail test_area_lim (#38212)

arw2019 committed:
BUG in Series.interpolate: limit_area/limit_direction kwargs with method="pad"/"bfill" have no effect (#38106)

roberthdevries committed:
DEPR: Deprecate using `xlrd` engine for read_excel (#35029)

erictleung committed:
DOC: fix redirects to user guide in README (#38189)

jbrockmendel committed:
REF: IntervalIndex.intersection match pattern in other intersection methods (#38190)

jbrockmendel committed:
BUG: name retention in Index.intersection (#38111)

phofl committed:
CLN: improve is_unique check in Index.intersection (#38154)

phofl committed:
BUG: Avoid duplicates in DatetimeIndex.intersection (#38196)

jbrockmendel committed:
REF: implement _should_compare (#38105)

ooojpeg committed:
DOC: Clarify development environment creation for documentation changes (#38194)

Abhi-H committed:
Using os.PathLike instead of pathlib.Path (#37979) (#38018)

jbrockmendel committed:
REF: implement _ea_wrap_cython_operation (#38162)

GYHHAHA committed:
BUG: unexpected merge_ordered results caused by wrongly groupby (#38170)

ivanovmg committed:
TST: add comment that plotting tests are slow (#38171)

ooojpeg committed:
DOC: Fixed to_pickle compression argument list 'gzip' > 'gz' (#38192)

vikramaditya91 committed:
Remove duplicate entry on series.rst (#38199)

jusexton committed:
TST: Added dataframe constructor tests confirming order is preserved with standard Python dicts (#38033) (#38206)

Qbiwan committed:
change dict to literal (#38207)

Qbiwan committed:
Rewrite dict literal for files in tests/groupby & ./pandas/tests/dtypes  (#38205)

AbdulMAbdi committed:
Rewrite dict literal for files in ./pandas/tests/tools/ and  ./pandas/tests/resample/  (#38203)

jorisvandenbossche committed:
CI: pin pip to 20.2 on numpy-dev build (#38228)

jbrockmendel committed:
BUG: name attr in RangeIndex.intersection (#38197)

jbrockmendel committed:
TST: tighten testing assertions (#38223)

xhochy committed:
Update Dockerfile to use conda-forge compilers (#38219)

ivanovmg committed:
TST: add cases for nearest_workday and before_nearest_workday (#38215)

jbrockmendel committed:
Retain views with listlike indexers setitem (#38204)

jbrockmendel committed:
BUG: DataFrame.idxmin/idxmax with mixed dtypes (#38195)

jbrockmendel committed:
BUG: groupby.rank for dt64tz, period dtypes (#38187)

arw2019 committed:
ENH: support nanoseconds in Period constructor (#38175)

jbrockmendel committed:
REF: avoid try/except in wrapping in cython_agg_blocks (#38164)

GYHHAHA committed:
BUG: array-like quantile fails on column groupby (#38173)

phofl committed:
BUG: DataFrame.loc returning empty result with negative stepsize for MultiIndex (#38071)

phofl committed:
BUG: Index.drop raising Error when Index has duplicates (#38070)

simonjayhawkins committed:
REGR: unstack on 'int' dtype prevent fillna to work (#38244)

ssortman committed:
BUG: assert_frame_equal exception for datetime #37609 (#38157)

jbrockmendel committed:
REF: casting in _python_agg_general (#38235)

jbrockmendel committed:
REF: consolidate result-casting in _cython_operation (#38237)

jbrockmendel committed:
restore (#38238)

ivanovmg committed:
TST: setitem in presence of duplicate cols (#38241)

arw2019 committed:
delete unused import (#38252)

jbrockmendel committed:
PERF: #37462 (#38248)

arw2019 committed:
TST: string representation of IntervalIndex (#38242)

jbrockmendel committed:
REF: simplify _cython_operation return  (#38253)

jbrockmendel committed:
TST: split unfocused groupby test (#38255)

jbrockmendel committed:
BUG: CategoricalDtype.__repr__ with RangeIndex categories (#38245)

* BUG: CategoricalDtype.__repr__ with RangeIndex categories

* fix broken tests

* fixup

mavismonica committed:
Rewrite dict call to literals (#38265)

vmdhhh committed:
C408 rewrite dict call to dict literals (#38263)

jbrockmendel committed:
CLN: assorted cleanups (#38275)

ZhihuiChen0903 committed:
The error message for unsupported index on shift() could be more explicit (#38176)

* Update base.py

* Update base.py

* Update base.py

* Update base.py

* Update common.py

* Update common.py

* Update common.py

* Update test_analytics.py

* Update base.py

* Update base.py

* Update common.py

* Update test_analytics.py

* Update base.py

* Update common.py

* Update common.py

* Update common.py

jbrockmendel committed:
BUG: name retention in Index.difference corner cases (#38268)

* BUG: name retention in Index.difference corner cases

* whatsnew

jbrockmendel committed:
TST: fixturize unfocused test (#38284)

jbrockmendel committed:
REF: de-duplicate intersection methods (#38283)

jbrockmendel committed:
ENH: implement _should_compare/_is_comparable_dtype for all Index subclasses (#38251)

UrielMaD committed:
pandas/core/strings/accesory.py Unnecessary dict call - rewrite as a literal #38138 (#38280)

joao-zanutto committed:
Removing dict calls (#38276)

gerardjorgensen committed:
C408 fix (#38260)

phofl committed:
BUG: Validate drop_duplicates ignore_index argument for bool (#38277)

jbrockmendel committed:
BUG: isin numeric vs string (#38279)

jbrockmendel committed:
BUG: DataFrame.apply with axis=1 and EA dtype (#38272)

twoertwein committed:
ENH: context-manager for chunksize/iterator-reader (#38225)

UrielMaD committed:
New dict call fixings #38138 (#38298)

jorisvandenbossche committed:
REF: move common arithmetic ops from Integer/FloatingArray to NumericArray (#38290)

jbrockmendel committed:
REF: implement Groupby idxmin, idxmax without fallback (#38264)

GYHHAHA committed:
BUG: groupby.apply on the NaN group drops values with original axes return (#38257)

fmaguire committed:
Fix typo in melt value_name futurewarning (#38301)

jbrockmendel committed:
TYP: @final in groupby (#38200)

* TYP: @final in groupby

* TYP: @final in groupby

* remove comment

* indent docstring, @final

ma3da committed:
BUG: Inserting array of same size with Series.loc raises ValueError (#38266)

jbrockmendel committed:
BUG: IntervalIndex.union with mismatched dtypes both empty (#38282)

jbrockmendel committed:
TST: tighten xfails (#38309)

* TST: tighten xfails

* troubleshoot

rhshadrach committed:
REGR: Groupby first/last/nth treats None as an observation (#38330)

* BUG: Groupby first/last/nth treats None as an observation

* Reverted test changes, whatsnew

* Reverted test changes, whatsnew

* Remove nth from whatsnew

simonjayhawkins committed:
REGR: Fix Index construction from Sparse["datetime64[ns]"] (#38332)

simonjayhawkins committed:
DOC: 1.1.5 release date (#38318)

jbrockmendel committed:
CLN: dtypes.cast (#38327)

fangchenli committed:
CLN: fix C408 #38138 (#38320)

Co-authored-by: VirosaLi <2EkF8qUgpNkj>

Arun12121 committed:
Removed unnecessary dict calls (#38319)

jbrockmendel committed:
TST: avoid skipping in frame arith test (#38310)

jbrockmendel committed:
TST: share transform tests (#38304)

rkc007 committed:
BUG: Fixes plotting with nullable integers (#32073) (#38014)

jbrockmendel committed:
CLN: remove now-unnecesary coerce_to_dtypes (#38321)

jbrockmendel committed:
TYP/BUG: fix passing incorrect type to interpolate_1d (#38306)

phofl committed:
BUG: Clean intersection and fix resulting names when MultiIndex are equal (#38302)

jbrockmendel committed:
REF: simplify casting of scalars to array (#38313)

jbrockmendel committed:
REF: avoid accessing private iloc methods (#38337)

jbrockmendel committed:
TST: revert too-ambitious (#38345)

jbrockmendel committed:
TYP: datetimelike (#38307)

arw2019 committed:
ENH: add "schema" kwarg to io.sql.get_schema method (#38146)

* Added schema kwarg and test

* Added type hinting and marked GH issue in test case

* fixed pep8 issues

* Fixed sorting on import

* move comment

* styling

* review comment

* docstring

Co-authored-by: Gregory Livschitz <greg15001@gmail.com>

jbrockmendel committed:
PERF: isin casting integers to object (#38346)

rhshadrach committed:
DEPR: xlwt for writing excel files (#38317)

jbrockmendel committed:
ENH: use correct dtype in groupby cython ops when it is known (without try/except) (#38291)

Co-authored-by: Joris Van den Bossche <jorisvandenbossche@gmail.com>

Commit:
RLS: 1.2.0rc0

simonjayhawkins committed:
Start 1.3.0

simonjayhawkins committed:
DOC: minor fixes to 1.2 release notes (#38361)

simonjayhawkins committed:
CI: activate azure pipelines/github actions on 1.2.x (#38360)

arw2019 committed:
TST: a weird bug with numpy + DataFrame with duplicate columns (#38354)

jbrockmendel committed:
REF: call _ensure_valid_index inside _sanitize_column (#38349)

simonjayhawkins committed:
DOC: Start v1.3.0 release notes (#38359)

jbrockmendel committed:
REF: use lighter-weight casting function (#38322)

erictleung committed:
DOC: add examples to insert and update generally (#37690)

* DOC: add examples to insert and update generally

This commit mainly adds examples on usage. This commit also fills in
information suggested by `validate_docstrings.py` script.

* DOC: remove inferred insert parameter descriptions

jbrockmendel committed:
BUG: item_cache invalidation (#38351)

WillAyd committed:
Removed unused strict struct member (#38348)

fangchenli committed:
CLN: use with instead of try finally in io/stata.py (#38326)

Co-authored-by: VirosaLi <2EkF8qUgpNkj>

fangchenli committed:
DOC: sync xarray version (#38369)

Co-authored-by: VirosaLi <2EkF8qUgpNkj>

jbrockmendel committed:
REF: implement Index._get_indexer (#38308)

JoseNavy committed:
TST: add message matches to pytest.raises in various tests GH30999 (#38350)

UrielMaD committed:
CLN: C408 Unnecessary dict call - rewrite as a literal #38138 (#38383)

* last dict fixings

* last dict fixings

* last dict fixings

* last dict fixings

jbrockmendel committed:
typo fixup (#38374)

fangchenli committed:
CLN: fix C408 #38138 (#38385)

Co-authored-by: VirosaLi <2EkF8qUgpNkj>

jbrockmendel committed:
CLN: lighter-weight casting (#38377)

imaskm committed:
Corrected the grammar in Exception (#38216)

* Corrected the grammar in Exception

Exception on different arrays' length corrected

* Updated error message in tests file as well

Co-authored-by: Ashwani <ashwani.ashwani@external.thalesgroup.com>

jbrockmendel committed:
REF: consolidate NDFrame._iset_item, _set_item with DataFrame methods (#38389)

WillAyd committed:
Document Tips for Debugging C Extensions (#35100)

luckyvs1 committed:
Merge branch 'add-test-for-describe' of https://github.com/luckyvs1/pandas into add-test-for-describe

luckyvs1 committed:
Use Dataframe to reference pandas dataframe

mroeschke commented:
Closing in favor of https://github.com/pandas-dev/pandas/pull/38403

luckyvs1 commented:
Note for reference: Reason the diff turned out to be the way it is was due to the forked branch not being up to date with the latest, to fix this at this point do the following with the rebase applied can do the following:

1. `git flog` to find the commit before the rebase
2. `git reset --hard HEAD@{#}` to reset the branch
3. `git push --force` force update to fix the commit diff

From here, we can update the forked master first then rebase off that branch:
1. While on the master branch on the forked repo use the following to update the forked master branch: https://stackoverflow.com/questions/7244321/how-do-i-update-a-github-forked-repository
2. Then switch over to the branch `add-test-for-describe` and rebase with the forked master and the resolve any conflicts as needed the commit history will be a clean rebase as intended
